### PR TITLE
ETB#92 - Init module dev

### PR DESCRIPTION
**Pull request by jrachucki-n7s**
Init module added.

### CODE REVIEW

File: src/Init/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,22 @@
+
+project(Samv71FreeRtos VERSION 1.0.0 LANGUAGES C)
```
Comment (Glamhoth) [inline src/Init/CMakeLists.txt]: This should be renamed to something like `Samv71RuntimeInit`

File: src/Init/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+add_library(Samv71FreeRtosInit STATIC)
```
Comment (Glamhoth) [inline src/Init/CMakeLists.txt]: This should be renamed to something like Samv71RuntimeInit

File: src/Init/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+                SAMV71::Nvic
+                SAMV71::Pio
+                SAMV71::Pmc
+                SAMV71::Uart
+                SAMV71::Wdt
+                SAMV71::Sdramc
+                SAMV71::FreeRTOS)
+
+set_target_properties(Samv71FreeRtosInit PROPERTIES OUTPUT_NAME "init")
+add_library(Samv71FreeRtos::Init ALIAS Samv71FreeRtosInit)
```
Comment (Glamhoth) [inline src/Init/CMakeLists.txt]: This should be renamed to something like `SAMV71::Runtime::Init`

File: src/FreeRTOS/FreeRTOSConfig.h [hunk 4]
```diff
@@ -65,6 +65,11 @@
 #define configUSE_APPLICATION_TASK_TAG          0
 #define configUSE_COUNTING_SEMAPHORES           1
 
+#define configAPPLICATION_ALLOCATED_HEAP        1
```
Comment (Glamhoth) [inline src/FreeRTOS/FreeRTOSConfig.h:69]: maybe to `#ifndef`?

File: src/Init/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    PRIVATE     SAMV71::Startup
+                SAMV71::Fpu
+                SAMV71::Nvic
+                SAMV71::Pio
+                SAMV71::Pmc
+                SAMV71::Uart
+                SAMV71::Wdt
+                SAMV71::Sdramc
+                SAMV71::FreeRTOS)
+
```
Comment (Glamhoth) [inline src/Init/CMakeLists.txt:20]: This module should be formatted by `add_format_target`

File: src/app/main.c [hunk 6]
```diff
@@ -1,173 +1,113 @@
...
-{
-    int errCode = 0;
+    /* Initialise xNextWakeTime - this only needs to be done once. */
+    xNextWakeTime = xTaskGetTickCount();
 
-    Pmc_Config pmcConfig = SystemConfig_getPmcDefaultConfig();
-    Pmc_setConfig(&pmcConfig, &errCode);
+    for (;;) {
+        /* Send to the queue - causing the queue receive task to unblock and
+            toggle the LED.  0 is used as the block time so the sending operation
```
Comment (Glamhoth) [inline src/app/main.c]: LED?

File: src/app/main.c [hunk 7]
```diff
@@ -1,173 +1,113 @@
...
+            will not block - it shouldn't need to block as the queue should always
+            be empty at this point in the code. */
+        /* Place this task in the blocked state until it is time to run again. */
+        // vTaskDelayUntil(&xNextWakeTime, TASK1_FREQUENCY);
```
Comment (Glamhoth) [inline src/app/main.c]: dead code?

File: src/app/main.c [hunk 8]
```diff
@@ -1,173 +1,113 @@
...
-    Uart_write(&uart, 0x6f, UART_TIMEOUT, NULL);
-    Uart_write(&uart, 0x72, UART_TIMEOUT, NULL);
-    Uart_write(&uart, 0x6c, UART_TIMEOUT, NULL);
-    Uart_write(&uart, 0x64, UART_TIMEOUT, NULL);
-    Uart_write(&uart, 0x21, UART_TIMEOUT, NULL);
+        ulValueToSend++;
+        if (ulValueToSend > 'z') {
+            ulValueToSend = 'A';
+        }
+        // /* Place this task in the blocked state until it is time to run again. */
```
Comment (Glamhoth) [inline src/app/main.c]: Double comment

File: src/FreeRTOS/FreeRTOSConfig.h (hunk 4 above)
Comment (Glamhoth) [reply src/FreeRTOS/FreeRTOSConfig.h:69]: I mean - use this define only if user didn't set it himself

File: src/Init/CMakeLists.txt (hunk 1 above)
Comment (jrachucki-n7s) [reply src/Init/CMakeLists.txt]: fixed in new commit

File: src/Init/CMakeLists.txt (hunk 2 above)
Comment (jrachucki-n7s) [reply src/Init/CMakeLists.txt]: fixed in new commit

File: src/Init/CMakeLists.txt (hunk 3 above)
Comment (jrachucki-n7s) [reply src/Init/CMakeLists.txt]: fixed in new commit

File: src/FreeRTOS/FreeRTOSConfig.h (hunk 4 above)
Comment (jrachucki-n7s) [reply src/FreeRTOS/FreeRTOSConfig.h:69]: fixed in new commit

File: src/app/main.c (hunk 6 above)
Comment (jrachucki-n7s) [reply src/app/main.c]: fixed in new commit

File: src/app/main.c (hunk 7 above)
Comment (jrachucki-n7s) [reply src/app/main.c]: fixed in new commit

File: src/app/main.c (hunk 8 above)
Comment (jrachucki-n7s) [reply src/app/main.c]: fixed in new commit

**Review by Glamhoth — DISMISSED** [review]

File: src/Init/Init.c [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+        .driveStrength = 0,
+        .lowPowerEnableTimeout = Sdramc_LowPowerEnableTimeout_LpLastXfer128,
+        .deviceType = Sdramc_MemoryDeviceType_Sdram,
+        .loadModeRegisterCommandToActiveOrRefreshCommand = 2,
+        .supportUnalignedAccess = Sdramc_SupportUnalignedAccess_Supported,
+        .isRefreshErrorStatusInterruptEnabled = false,
+    };
+
+    Sdramc_setConfig(sdramc, &IS42S16100H_SDRAM_CONFIG);
+}
```
Comment (Lurkerpas) [inline src/Init/Init.c]: newline

File: src/Init/Init.c [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+Init_setup_nvic(void)
+{
+    Nvic_enableInterrupt(Nvic_Irq_Uart4);
+    Nvic_setInterruptPriority(Nvic_Irq_Uart4, 0);
+}
+
+static void
+Init_setup_uart(void)
+{
+    Uart_Config uartConfig = { .isTxEnabled = true,
```
Comment (Lurkerpas) [inline src/Init/Init.c:148]: I understand that this is for testing... but I would prefer to put here a TODO indicating that UART should be removed from here. All UART initialization should be done by HAL, which may additionally initialize USART for debug output.

Comment (jrachucki-n7s) [reply src/Init/Init.c:148]: Sure, I wanted to change that after implementing HAL. I will add TODO.

**Review by Glamhoth — APPROVED** [review]

**Pull request merged**